### ISSUE TITLE
Allow customization of namespace entrypoint file name.

### ISSUE THREAD

**Issue by jordwalke** [ocaml/dune#2648]
**Feature request:** Allow a library:

```
(library
  (name MyPackageMyLibrary)
  (public_name my-package.my-library)
)
```

To have an entry point file name other than `MyPackageMyLibrary.ml/re`.

**Why?**

@prometheansacrifice and I discovered a set of use cases that this feature unlocks. I think these use cases could end up benefiting the overall integrity of the ocaml ecosystem because it would help avoid namespace collisions.

- We want to encourage people to avoid collisions in namespaces.
- Therefore we would encourage people to make their library namespaces pretty long, including their package name as the prefix.
- Then when consuming them, they should alias them to something shorter.
- But when developing these long-named-libraries, they want to use the more shorter file name.
- If we don't let them use the shorter file name with a longer namespace, they're more likely to just use a shorter (more collision-prone) namespace, instead of making their entrypoint file name longer named - and the result is more collisions in the ecosystem.

**Comment by jordwalke** [comment]
With this feature added, there are other things to be done that make longer namespaces more manageable - and this feature doesn't propose solutions to those. But this is one feature that could unblock experimentation with those other patterns/conventions which Dune could eventually encode as a best practice.

**Comment by jordwalke** [comment]
Another feature that (in conjunction with this request) would make longer namespaces tolerable: https://github.com/ocaml/dune/issues/2649

**Comment by diml** [comment]
Just to make sure I understand the proposal correctly, is the idea that instead of writing a file `myPackageMyLibrary.ml` you would instead write a file `myLibrary.ml` but from the outside, modules of the library would still be accessed via `MyPackageMyLibrary.X`?

**Comment by jordwalke** [comment]
@diml Correct. This proposal is a bite sized feature that still requires you to access the longer module namespace from the outside. The linked issue I created then proposes some facilities for allowing you to rename them to something shorter at consumption time. I consider them separate proposals because there are other reasons why you might want the entrypoint module name to differ from the namespace name (for example, you main want a `(main init.ml)` etc.

**Comment by diml** [comment]
Ok, this proposal seems fine to me. The only thing to make sure of is that this works well with the tooling, i.e. merlin and odoc.

Do you want to give a shot at implementing this feature?

**Comment by jordwalke** [comment]
I think we can take a look at implementing, but we want to leave this open for more feedback and implement it in user space to see if this is as useful as we suspect it will be (I am optimistic).

**Comment by diml** [comment]
Sounds good. On the technical side, you need to edit the type `Library.t` and function `Library.decode` in `src/dune/dune_file.ml` to add a `main` field. You then need to adapt the last branch of the function `Library.main_module_name` and return the name set by the user instead of the inferred one. Then, if everything is already properly wired up it should just work. Once this works, you'll need to add a test in `test/blackbox-tests/test-cases` following the format of the other tests and add some documentation to the manual in `doc/dune-files.rst`.

**Comment by mjambon** [comment]
I like this proposal because it allows the user to decouple the name of the library (external) from the name of the entry point (internal). The developer of a multi-library project could use `Main.ml` systematically for the entry point.

**Comment by diml** [comment]
Indeed. For the case where the entry point just contains the list of modules, we could also slightly extend this proposal so that one doesn't have to write a boring `main.ml` file. For instance by writing `(main (implicit))`

**Comment by jordwalke** [comment]
Thank you for the pointers, @diml - very helpful.
